test/distributed/checkpoint/e2e/test_fine_tuning.py

```diff
@@ -0,0 +1,192 @@
+        x = self.relu(self.pretrain(batch))
+        x = self.relu(self.layer1(x))
+        x = self.relu(self.layer2(x))
+        x = self.relu(self.layer3(x))
+        return x
+
+
+class TestFineTuning(FSDPTest):
+    @property
+    def world_size(self) -> int:
+        return min(4, torch.cuda.device_count())
+
+    def pretrain(self, pretrain_dir: str) -> None:
+        model = PreTrainedModel().cuda()
+        model = FSDP(model)
+        optim = torch.optim.Adam(model.parameters(), lr=1e-3)
+
+        # Trainining
+        for i in range(3):
+            batch = torch.rand(32, DIM, device="cuda")
+            loss = model(batch).sum()
+            loss.backward()
+            optim.step()
+            optim.zero_grad()
+
+        # Save state_dict
+        model_state_dict, optim_state_dict = get_state_dict(model, optimizers=optim)
+        saved_state_dict = {"model": model_state_dict, "optim": optim_state_dict}
+        dist_cp.save_state_dict(
+            state_dict=saved_state_dict,
+            storage_writer=dist_cp.FileSystemWriter(pretrain_dir),
+        )
+
+    def finetune(self, pretrain_dir: str, finetune_dir: str) -> None:
+        model = FineTuningModel().cuda()
+        # TODO: make the parallelism more complicated, e.g., using 2D + DDP.
+        model = FSDP(model, use_orig_params=True)
+        optim = torch.optim.Adam(model.parameters(), lr=1e-3)
+
+        # Simulate that the fine tuning restart after 3 iterations
+        for i in range(2):
+            # Load pretrain submodules checkpoint
+            pretrain_state_dict, _ = get_state_dict(
+                model,
+                submodules={model.pretrain},
+                options=StateDictOptions(keep_submodule_prefixes=False)
+            )
+            dist_cp.load_state_dict(
+                {"model": pretrain_state_dict},
+                storage_reader=dist_cp.FileSystemReader(pretrain_dir),
+            )
+            set_state_dict(
+                model,
+                model_state_dict={model.pretrain: pretrain_state_dict},
+                options=StateDictOptions(strict=False),
+            )
+
+            try:
+                # Load training submodules checkpoint
+                model_state_dict, optim_state_dict = get_state_dict(
+                    model,
+                    optimizers=optim,
+                    options=StateDictOptions(ignore_frozen_params=True),
+                )
+                dist_cp.load_state_dict(
+                    {"model": model_state_dict, "optim": optim_state_dict},
+                    storage_reader=dist_cp.FileSystemReader(pretrain_dir),
+                )
+                set_state_dict(
+                    model,
+                    optimizers=optim,
+                    model_state_dict=model_state_dict,
+                    optim_state_dict=optim_state_dict,
+                    options=StateDictOptions(strict=False),
+                )
+            except KeyError:
```

Comment:
I don't quite understand the explanation below - why do we hit a KeyError here on non-zero iteration? Shouldn't we be able to load the finetuning part in the middle of training?

Reply:
Oh nvm, this is because we haven't done any finetuning yet, so there isn't a finetune state_dict to load. Unsure if there's a way to make this more clear.

Reply:
Correct, DCP does not have an API to check, at this moment. So we have to rely on try/except.